File: internal/crypto/crypto.go
// Package crypto provides WireGuard-style encryption for tunnel traffic.
//
// Implements X25519 key exchange and ChaCha20-Poly1305/AES-256-GCM frame encryption
// for securing data in transit through Burrow tunnels.
package crypto

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"math"
	"sync"
	"sync/atomic"
	"time"

	"golang.org/x/crypto/chacha20poly1305"
	"golang.org/x/crypto/curve25519"
	"golang.org/x/crypto/hkdf"

	"crypto/sha256"
)

// CipherSuite identifies the AEAD cipher used for frame encryption.
type CipherSuite int

const (
	// ChaCha20Poly1305 is the default cipher suite (WireGuard-style).
	ChaCha20Poly1305 CipherSuite = iota
	// AES256GCM uses AES-256 in GCM mode.
	AES256GCM
)

const (
	// KeySize is the size of X25519 keys and derived symmetric keys (32 bytes).
	KeySize = 32
	// NonceSize is 12 bytes for both ChaCha20-Poly1305 and AES-256-GCM.
	NonceSize = 12
	// CounterSize is 4 bytes for the frame counter prefix.
	CounterSize = 4
	// FrameHeaderSize is CounterSize + NonceSize.
	FrameHeaderSize = CounterSize + NonceSize
	// MaxFrameSize limits individual encrypted frames to 64KB.
	MaxFrameSize = 65536
	// DefaultRotationInterval is the key rotation interval (1 hour).
	DefaultRotationInterval = time.Hour
)

var (
	// ErrPeerKeyNotSet is returned when attempting crypto operations without a peer key.
	ErrPeerKeyNotSet = errors.New("peer public key not set")
	// ErrFrameTooShort is returned when an encrypted frame is smaller than the header.
	ErrFrameTooShort = errors.New("frame too short")
	// ErrDecryptFailed is returned when AEAD decryption fails (tampered data).
	ErrDecryptFailed = errors.New("decryption failed: invalid ciphertext or tag")
	// ErrReplayDetected is returned when a decrypted frame has an already-seen counter.
	ErrReplayDetected = errors.New("replay detected: counter not advancing")
	// ErrCounterExhausted is returned when the send counter exceeds uint32 max.
	ErrCounterExhausted = errors.New("counter exhausted: rotate key before sending more data")
)

// KeyPair holds an X25519 private/public key pair.
type KeyPair struct {
	Private [KeySize]byte
	Public  [KeySize]byte
}

// GenerateKeyPair creates a new X25519 key pair.
func GenerateKeyPair() (*KeyPair, error) {
	kp := &KeyPair{}
	if _, err := rand.Read(kp.Private[:]); err != nil {
		return nil, fmt.Errorf("generate key pair: %w", err)
	}
	// Clamp private key per X25519 spec
	kp.Private[0] &= 248
	kp.Private[31] &= 127
	kp.Private[31] |= 64

	pub, err := curve25519.X25519(kp.Private[:], curve25519.Basepoint)
	if err != nil {
		return nil, fmt.Errorf("compute public key: %w", err)
	}
	copy(kp.Public[:], pub)
	return kp, nil
}

// DeriveSharedKey performs X25519 ECDH and derives a 256-bit key via HKDF-SHA256.
func DeriveSharedKey(privateKey, peerPublicKey [KeySize]byte) ([KeySize]byte, error) {
	var derived [KeySize]byte

	shared, err := curve25519.X25519(privateKey[:], peerPublicKey[:])
	if err != nil {
		return derived, fmt.Errorf("ECDH exchange: %w", err)
	}

	hkdfReader := hkdf.New(sha256.New, shared, nil, []byte("burrow-tunnel-v1"))
	if _, err := io.ReadFull(hkdfReader, derived[:]); err != nil {
		return derived, fmt.Errorf("HKDF derive: %w", err)
	}
	return derived, nil
}

// newAEAD creates an AEAD cipher from the given key and suite.
func newAEAD(key [KeySize]byte, suite CipherSuite) (cipher.AEAD, error) {
	switch suite {
	case ChaCha20Poly1305:
		return chacha20poly1305.New(key[:])
	case AES256GCM:
		block, err := aes.NewCipher(key[:])
		if err != nil {
			return nil, err
		}
		return cipher.NewGCM(block)
	default:
		return nil, fmt.Errorf("unsupported cipher suite: %d", suite)
	}
}

// buildNonce creates a 12-byte nonce: 4 zero bytes + 4-byte BE counter + 4 random bytes.
func buildNonce(counter uint32) ([NonceSize]byte, error) {
	var nonce [NonceSize]byte
	// First 4 bytes: zero padding
	// Next 4 bytes: big-endian counter
	binary.BigEndian.PutUint32(nonce[4:8], counter)
	// Last 4 bytes: random
	if _, err := rand.Read(nonce[8:]); err != nil {
		return nonce, err
	}
	return nonce, nil
}

// EncryptFrame encrypts plaintext into a framed message:
// counter(4) || nonce(12) || ciphertext+tag
func EncryptFrame(key [KeySize]byte, plaintext []byte, counter uint32, suite CipherSuite, aad []byte) ([]byte, error) {
	aead, err := newAEAD(key, suite)
	if err != nil {
		return nil, fmt.Errorf("create AEAD: %w", err)
	}

	nonce, err := buildNonce(counter)
	if err != nil {
		return nil, fmt.Errorf("build nonce: %w", err)
	}

	frame := make([]byte, FrameHeaderSize+len(plaintext)+aead.Overhead())
	binary.BigEndian.PutUint32(frame[:CounterSize], counter)
	copy(frame[CounterSize:FrameHeaderSize], nonce[:])

	ciphertext := aead.Seal(frame[FrameHeaderSize:FrameHeaderSize], nonce[:], plaintext, aad)
	return frame[:FrameHeaderSize+len(ciphertext)], nil
}

// DecryptFrame decrypts a framed message, returning the counter and plaintext.
func DecryptFrame(key [KeySize]byte, frame []byte, suite CipherSuite, aad []byte) (uint32, []byte, error) {
	if len(frame) < FrameHeaderSize {
		return 0, nil, ErrFrameTooShort
	}

	aead, err := newAEAD(key, suite)
	if err != nil {
		return 0, nil, fmt.Errorf("create AEAD: %w", err)
	}

	counter := binary.BigEndian.Uint32(frame[:CounterSize])
	var nonce [NonceSize]byte
	copy(nonce[:], frame[CounterSize:FrameHeaderSize])
	ciphertext := frame[FrameHeaderSize:]

	plaintext, err := aead.Open(nil, nonce[:], ciphertext, aad)
	if err != nil {
		return 0, nil, ErrDecryptFailed
	}
	return counter, plaintext, nil
}

// Session manages an encrypted tunnel session with counter tracking and key rotation.
type Session struct {
	localKey    *KeyPair
	peerPublic  [KeySize]byte
	sharedKey   [KeySize]byte
	suite       CipherSuite
	hasPeer     bool
	hasShared   bool
	sendCounter atomic.Uint64
	recvCounter atomic.Uint64
	keyCreated  time.Time
	rotationInt time.Duration
	mu          sync.RWMutex
}

// NewSession creates a new crypto session with a fresh key pair.
func NewSession(suite CipherSuite) (*Session, error) {
	kp, err := GenerateKeyPair()
	if err != nil {
		return nil, err
	}
	return &Session{
		localKey:    kp,
		suite:       suite,
		keyCreated:  time.Now(),
		rotationInt: DefaultRotationInterval,
	}, nil
}

// LocalPublicKey returns this session's public key for sharing with the peer.
func (s *Session) LocalPublicKey() [KeySize]byte {
	return s.localKey.Public
}

// SetPeerPublicKey sets the peer's public key and derives the shared secret.
func (s *Session) SetPeerPublicKey(peerPub [KeySize]byte) error {
	s.mu.Lock()
	defer s.mu.Unlock()

	s.peerPublic = peerPub
	s.hasPeer = true

	shared, err := DeriveSharedKey(s.localKey.Private, peerPub)
	if err != nil {
		return err
	}
	s.sharedKey = shared
	s.hasShared = true
	s.keyCreated = time.Now()
	s.sendCounter.Store(0)
	s.recvCounter.Store(0)
	return nil
}

// Encrypt encrypts plaintext with the session's shared key, auto-incrementing the counter.
func (s *Session) Encrypt(plaintext []byte, aad []byte) ([]byte, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if !s.hasShared {
		return nil, ErrPeerKeyNotSet
	}

	cnt := s.sendCounter.Add(1) - 1
	if cnt > math.MaxUint32 {
		return nil, ErrCounterExhausted
	}
	counter := uint32(cnt)
	return EncryptFrame(s.sharedKey, plaintext, counter, s.suite, aad)
}

// Decrypt decrypts a frame with the session's shared key.
func (s *Session) Decrypt(frame []byte, aad []byte) ([]byte, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	if !s.hasShared {
		return nil, ErrPeerKeyNotSet
	}

	counter, plaintext, err := DecryptFrame(s.sharedKey, frame, s.suite, aad)
	if err != nil {
		return nil, err
	}

	// Enforce strictly advancing counters for anti-replay.
	next := uint64(counter) + 1
	for {
		current := s.recvCounter.Load()
		if next <= current {
			return nil, ErrReplayDetected
		}
		if s.recvCounter.CompareAndSwap(current, next) {
			break
		}
	}
	return plaintext, nil
}

// NeedsRotation returns true if the session key should be rotated.
func (s *Session) NeedsRotation() bool {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return time.Since(s.keyCreated) > s.rotationInt
}

// RotateKey generates a new key pair and resets counters. Returns new public key.
func (s *Session) RotateKey() ([KeySize]byte, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	kp, err := GenerateKeyPair()
	if err != nil {
		return [KeySize]byte{}, err
	}
	s.localKey = kp
	s.hasShared = false
	s.sendCounter.Store(0)
	s.recvCounter.Store(0)
	return kp.Public, nil
}

// SendCounter returns the current send counter value.
func (s *Session) SendCounter() uint64 {
	return s.sendCounter.Load()
}

// RecvCounter returns the current recv counter value.
func (s *Session) RecvCounter() uint64 {
	return s.recvCounter.Load()
}

File: src/burrow/crypto.py
"""WireGuard-style encryption for tunnel traffic.

Provides X25519 key exchange and ChaCha20-Poly1305 / AES-256-GCM frame encryption.
"""

from __future__ import annotations

import os
import struct
import time
from dataclasses import dataclass, field
from enum import Enum

from cryptography.hazmat.primitives.asymmetric.x25519 import (
    X25519PrivateKey,
    X25519PublicKey,
)
from cryptography.hazmat.primitives.ciphers.aead import AESGCM, ChaCha20Poly1305
from cryptography.hazmat.primitives.hashes import SHA256
from cryptography.hazmat.primitives.kdf.hkdf import HKDF


class CipherSuite(str, Enum):
    """Supported cipher suites for tunnel encryption."""

    CHACHA20_POLY1305 = "chacha20-poly1305"
    AES_256_GCM = "aes-256-gcm"


# Frame layout: [4-byte nonce_counter_be][12-byte nonce][N-byte ciphertext+tag]
NONCE_SIZE = 12
COUNTER_SIZE = 4
FRAME_HEADER_SIZE = COUNTER_SIZE + NONCE_SIZE


def generate_keypair() -> tuple[X25519PrivateKey, X25519PublicKey]:
    """Generate an X25519 key pair for ECDH key exchange."""
    private_key = X25519PrivateKey.generate()
    public_key = private_key.public_key()
    return private_key, public_key


def derive_shared_key(
    private_key: X25519PrivateKey,
    peer_public_key: X25519PublicKey,
    info: bytes = b"burrow-tunnel-v1",
) -> bytes:
    """Derive a 256-bit shared key from ECDH output using HKDF-SHA256."""
    shared_secret = private_key.exchange(peer_public_key)
    derived = HKDF(
        algorithm=SHA256(),
        length=32,
        salt=None,
        info=info,
    ).derive(shared_secret)
    return derived


def _build_nonce(counter: int) -> bytes:
    """Build a 12-byte nonce: 4 bytes of zero padding + 4-byte big-endian counter + 4 random bytes."""
    return struct.pack(">I", 0) + struct.pack(">I", counter) + os.urandom(4)


def encrypt_frame(
    key: bytes,
    plaintext: bytes,
    counter: int,
    cipher_suite: CipherSuite = CipherSuite.CHACHA20_POLY1305,
    aad: bytes | None = None,
) -> bytes:
    """Encrypt a data frame.

    Returns: counter_bytes(4) || nonce(12) || ciphertext_with_tag
    """
    nonce = _build_nonce(counter)
    if cipher_suite == CipherSuite.CHACHA20_POLY1305:
        cipher = ChaCha20Poly1305(key)
    else:
        cipher = AESGCM(key)  # type: ignore[assignment]
    ciphertext = cipher.encrypt(nonce, plaintext, aad)
    return struct.pack(">I", counter) + nonce + ciphertext


def decrypt_frame(
    key: bytes,
    frame: bytes,
    cipher_suite: CipherSuite = CipherSuite.CHACHA20_POLY1305,
    aad: bytes | None = None,
) -> tuple[int, bytes]:
    """Decrypt a data frame.

    Returns: (counter, plaintext)
    Raises: cryptography.exceptions.InvalidTag on tampered data.
    """
    if len(frame) < FRAME_HEADER_SIZE:
        raise ValueError(f"Frame too short: {len(frame)} < {FRAME_HEADER_SIZE}")
    counter = struct.unpack(">I", frame[:COUNTER_SIZE])[0]
    nonce = frame[COUNTER_SIZE:FRAME_HEADER_SIZE]
    ciphertext = frame[FRAME_HEADER_SIZE:]
    if cipher_suite == CipherSuite.CHACHA20_POLY1305:
        cipher = ChaCha20Poly1305(key)
    else:
        cipher = AESGCM(key)  # type: ignore[assignment]
    plaintext = cipher.decrypt(nonce, ciphertext, aad)
    return counter, plaintext


@dataclass
class CryptoSession:
    """Manages an encrypted tunnel session with key rotation support."""

    local_private: X25519PrivateKey = field(default_factory=X25519PrivateKey.generate)
    peer_public: X25519PublicKey | None = None
    cipher_suite: CipherSuite = CipherSuite.CHACHA20_POLY1305
    _shared_key: bytes | None = field(default=None, repr=False)
    _send_counter: int = field(default=0, repr=False)
    _recv_counter: int = field(default=0, repr=False)
    _key_created_at: float = field(default_factory=time.time, repr=False)
    _rotation_interval: float = field(default=3600.0, repr=False)  # 1 hour

    @property
    def local_public(self) -> X25519PublicKey:
        """Return our public key for sharing with peer."""
        return self.local_private.public_key()

    @property
    def shared_key(self) -> bytes:
        """Return the derived shared key, computing lazily."""
        if self._shared_key is None:
            if self.peer_public is None:
                raise ValueError("Peer public key not set — call set_peer_public first")
            self._shared_key = derive_shared_key(self.local_private, self.peer_public)
            self._key_created_at = time.time()
        return self._shared_key

    def set_peer_public(self, peer_pub: X25519PublicKey) -> None:
        """Set peer public key and derive shared secret."""
        self.peer_public = peer_pub
        self._shared_key = derive_shared_key(self.local_private, peer_pub)
        self._key_created_at = time.time()
        self._send_counter = 0
        self._recv_counter = 0

    def encrypt(self, plaintext: bytes, aad: bytes | None = None) -> bytes:
        """Encrypt plaintext, auto-incrementing the send counter."""
        frame = encrypt_frame(
            self.shared_key, plaintext, self._send_counter, self.cipher_suite, aad
        )
        self._send_counter += 1
        return frame

    def decrypt(self, frame: bytes, aad: bytes | None = None) -> bytes:
        """Decrypt frame, validating counter ordering."""
        counter, plaintext = decrypt_frame(
            self.shared_key, frame, self.cipher_suite, aad
        )
        self._recv_counter = max(self._recv_counter, counter + 1)
        return plaintext

    def needs_rotation(self) -> bool:
        """Check if the session key should be rotated."""
        return (time.time() - self._key_created_at) > self._rotation_interval

    def rotate_key(self) -> X25519PublicKey:
        """Generate a new keypair for rekeying. Returns new public key to send to peer."""
        self.local_private = X25519PrivateKey.generate()
        self._shared_key = None
        self._send_counter = 0
        self._recv_counter = 0
        return self.local_public

    @property
    def send_counter(self) -> int:
        """Current send counter value."""
        return self._send_counter

    @property
    def recv_counter(self) -> int:
        """Current recv counter value."""
        return self._recv_counter
